cli/ops/jupyter.rs

```diff
@@ -32,6 +43,97 @@ deno_core::extension!(deno_jupyter,
   },
 );
 
+pub struct CommChannel {
+  pub target_name: String,
+  pub sender: broadcast::Sender<(CommMsg, Vec<Bytes>)>,
+  pub receiver: broadcast::Receiver<(CommMsg, Vec<Bytes>)>,
+}
+
+#[derive(Clone, Default)]
+pub struct CommContainer(pub Arc<PlMutex<HashMap<String, CommChannel>>>);
+
+impl CommContainer {
+  pub fn create(
+    &mut self,
+    comm_id: &str,
+    target_name: &str,
+    // For pulling off the metadata and buffers
+    _msg: Option<&JupyterMessage>,
+  ) {
+    let mut container = self.0.lock();
+
+    // We will not replace existing comms
+    if container.contains_key(comm_id) {
+      return;
+    }
+
+    let (tx, rx) = broadcast::channel(16);
+    let comm_channel = CommChannel {
+      target_name: target_name.to_string(),
+      sender: tx,
+      receiver: rx,
+    };
+
+    container.insert(comm_id.to_string(), comm_channel);
+  }
+
+  pub fn comms(&self) -> HashMap<CommId, CommInfo> {
+    let container = self.0.lock();
+
+    container
+      .iter()
+      .map(|(comm_id, comm)| {
+        (
+          CommId(comm_id.to_string()),
+          CommInfo {
+            target_name: comm.target_name.clone(),
+          },
+        )
+      })
+      .collect()
+  }
+}
+
+#[op2(fast)]
+pub fn op_jupyter_comm_open(
+  state: &mut OpState,
+  #[string] comm_id: String,
+  #[string] target_name: String,
+) {
+  let container = state.borrow_mut::<CommContainer>();
+  container.create(&comm_id, &target_name, None);
+  eprintln!("created comm {} {}", comm_id, target_name);
+}
+
+#[op2(async)]
+#[serde]
+pub async fn op_jupyter_comm_recv(
+  state: Rc<RefCell<OpState>>,
+  #[string] comm_id: String,
+) -> (serde_json::Value, Vec<ToJsBuffer>) {
+  let mut receiver = {
+    let state = state.borrow();
+    let container = state.borrow::<CommContainer>();
+    let container = container.0.lock();
+    let maybe_comm = container.get(&comm_id);
+    let Some(comm) = maybe_comm else {
+      return (serde_json::Value::Null, vec![]);
+    };
+    comm.receiver.resubscribe()
```

Comment:
@manzt @rgbkrk if we go with `EventListener` approach for a comm then we can expect more than one consumer of the "comm". That means we need to use `tokio::sync::broadcast` channel which is buffered. We can go with a rather big number for the buffer like 1024 or 65536. If we expect only a single consumer then we could use an `mpsc` channel that is unbounded. Which one should we go with?

Reply:
For widgets, I think it will basically only be one comm listener. There is new comm created for each new widget instance.